### PR TITLE
[ONNX export] Support for image normalization in model graph.

### LINKED ISSUE #5958: [ONNX export] Support image normalization in graph
**Describe the feature**
Add option to include the image normalization preprocessing step in the ONNX graph.

**Motivation**
In order to export models to truly portable ONNX files, the preprocessing steps need to be included in the graph as much as possible. There is also a computational efficiency aspect, as normalization would happen on the GPU (e.g. see https://github.com/open-mmlab/mmdetection/issues/4488)

**Related resources**
A potential implementation would consist in passing the `normalize_cfg` to the `forward` method so that it can be used to normalize the image. I.e. adding the block below at https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/detectors/base.py#L169
```
            onnx_input = img[0]
            normalize_cfg = kwargs.get('normalize_cfg', None)
            if normalize_cfg:
                mean = torch.tensor(normalize_cfg['mean'])[..., None, None]
                std = torch.tensor(normalize_cfg['std'])[..., None, None]
                onnx_input = (img[0] - mean) / std
            return self.onnx_export(onnx_input, img_metas[0])
```

Then `tools/deployment/pytorch2onnx.py` can be modified accordingly to support the option.

**Additional context**
If approved, I'd be happy to open a PR with the changes above (and/or alternative/additional suggestions).

### PR DESCRIPTION
**Pull request by francescorubbo**
## Motivation

In order to export models to truly portable ONNX files, the preprocessing steps need to be included in the graph as much as possible.

## Modification

Add option to include the image normalization preprocessing step in the ONNX graph.
The normalization parameters are passed as `img_norm_cfg` via `img_meta` and applied in the forward method.

Fixes #5958

### CODE REVIEW

File: mmdet/models/detectors/base.py [hunk 1]
```diff
@@ -164,6 +164,12 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
         should be double nested (i.e.  List[Tensor], List[List[dict]]), with
         the outer list indicating test time augmentations.
         """
+        img_norm_cfg = kwargs.get('img_norm_cfg', None)
+        if img_norm_cfg:
+            mean = torch.tensor(img_norm_cfg['mean'])[None, ..., None, None]
+            std = torch.tensor(img_norm_cfg['std'])[None, ..., None, None]
```
Comment (ZwwWayne) [inline mmdet/models/detectors/base.py:170]: We suggest keeping this logic in lines between 173-175 to restrict its influence.

File: tools/deployment/pytorch2onnx.py [hunk 2]
```diff
@@ -280,6 +285,10 @@ def parse_args():
         help='Whether to export model without post process. Experimental '
         'option. We do not guarantee the correctness of the exported '
         'model.')
+    parser.add_argument(
+        '--normalize-in-graph',
+        action='store_true',
+        help='Whether to include image normalization in ONNX graph.')
```
Comment (RunningLeon) [inline tools/deployment/pytorch2onnx.py:291]: Could update the doc as well:
https://github.com/open-mmlab/mmdetection/blob/master/docs/tutorials/pytorch2onnx.md#usage

Comment (RunningLeon) [reply tools/deployment/pytorch2onnx.py:291]: If a user wants to deploy a model on TensorRT by PyTorch->ONNX->TensorRT, then, this optional feature would have an influence on `onnx2tensorrt.py` tool.

Comment (francescorubbo) [reply tools/deployment/pytorch2onnx.py:291]: I've updated the docs and added a `--skip-normalize` option to the `onnx2tensorrt.py` tool. However, I do not have MMCV built with the tensorRT support, so I am not able to test the change myself.

File: docs/tutorials/onnx2tensorrt.md [hunk 3]
```diff
@@ -50,6 +51,7 @@ Description of all arguments:
 - `--show`: Determines whether to show the outputs of the model. If not specified, it will be set to `False`.
 - `--verify`: Determines whether to verify the correctness of models between ONNXRuntime and TensorRT. If not specified, it will be set to `False`.
 - `--verbose`: Determines whether to print logging messages. It's useful for debugging. If not specified, it will be set to `False`.
+- `--skip-normalize`: Determins whether to skip normalization of the input image. If not specified, it will be set to `False`. This is useful when the normalization is already included in the ONNX graph.
```
Comment (RunningLeon) [inline docs/tutorials/onnx2tensorrt.md:54]: You could add note that this argument should be set to `True` only if `--normalize-in-graph` in `pytorch2onnx.py` is set to `True`.

File: tools/deployment/pytorch2onnx.py (hunk 2 above)
Comment (RunningLeon) [reply tools/deployment/pytorch2onnx.py:291]: You could follow the instructions in this doc [onnx2tensorrt](https://github.com/open-mmlab/mmdetection/blob/master/docs/tutorials/onnx2tensorrt.md) to test the logic.

Comment (francescorubbo) [reply tools/deployment/pytorch2onnx.py:291]: I'm developing on MacOS, so I can't install TensorRT. Regardless, I've tested that the option is properly picked up by the argparse and that the logic is correct in not including the `normalize_cfg` entry in `input_config` when the `--skip-normalize` is set to `True`.

**Comment by francescorubbo** [comment]
@RunningLeon are there additional changes needed for this? Are the tests actually failing (I can't seem to find the logs), or they are killed?

**Comment by RunningLeon** [comment]
> @RunningLeon are there additional changes needed for this? Are the tests actually failing (I can't seem to find the logs), or they are killed?

Seems ok to me.

**Comment by francescorubbo** [comment]
@RangiLyu what would be the appropriate place to add tests to improve test coverage? As far as I can tell, the tests in `tests/test_onnx` only test inference of model neck and head, not the end to end model that would be needed in this case.

**Comment by francescorubbo** [comment]
@RangiLyu thank you for getting back to me. I have added the test and fixed a bug that the test surfaced.
I suspect that we might still be lacking coverage, because the test only covers `mmdet/models/detectors/base.py`, not `mmdet/core/export/pytorch2onnx.py`. Any thought about that?

**Comment by ZwwWayne** [comment]
Hi @francescorubbo ,
Thanks for your kind contribution.
Since ONNX will be fully migrated to MMDeploy in the future and we are planning a more general design for the data pre-processing. This PR may not be merged and is thus closed.

**Comment by OpenMMLab-Assistant005** [comment]
Hi @francescorubbo ！We are grateful for your efforts in helping improve this open-source project during your personal time. 
Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
